File: mdix-java/src/main/java/com/midmanstudio/dixscript/DixScript.java
package com.midmanstudio.dixscript;

import com.midmanstudio.dixscript.internal.MdixNative;

/**
 * Static entry point for all DixScript operations.
 *
 * <pre>
 * // Load from file
 * try (Database db = DixScript.load("config.mdix")) {
 *     int port = db.getInt("server.port");
 * }
 *
 * // Load from string
 * try (Database db = DixScript.loadStr("@DATA( x = 42 )")) {
 *     int x = db.getInt("x");
 * }
 *
 * // Build
 * try (Builder b = DixScript.newBuilder()) {
 *     b.setString("name", "MyApp").setInt("version", 1);
 *     b.saveToFile("out.mdix");
 * }
 * </pre>
 */
public final class DixScript {

    private static final Converter CONVERTER = new Converter();

    private DixScript() {}

    // ── Metadata ──────────────────────────────────────────────────────────────

    public static String version() {
        return MdixNative.version();
    }

    // ── Loading ───────────────────────────────────────────────────────────────

    /**
     * Loads a .mdix file from disk.
     * @throws MdixException on parse or IO error
     */
    public static Database load(String path) {
        if (path == null || path.isEmpty())
            throw new MdixException(MdixException.Kind.INVALID_PATH, "path is null or empty");
        long h = MdixNative.load(path);
        if (h == 0)
            throw new MdixException(MdixException.Kind.IO_ERROR, "failed to load: " + path);
        return new Database(h);
    }

    /**
     * Loads .mdix content from a source string.
     * @throws MdixException on parse error
     */
    public static Database loadStr(String source) {
        if (source == null || source.isEmpty())
            throw new MdixException(MdixException.Kind.PARSE_ERROR, "source is null or empty");
        long h = MdixNative.loadStr(source);
        if (h == 0)
            throw new MdixException(MdixException.Kind.PARSE_ERROR, "failed to parse source");
        return new Database(h);
    }

    /**
     * Loads an encrypted .mdix.enc file using a key file.
     * Pass {@code null} for {@code keyPath} to auto-detect next to the enc file.
     */
    public static Database loadEncrypted(String encPath, String keyPath) {
        if (encPath == null || encPath.isEmpty())
            throw new MdixException(MdixException.Kind.INVALID_PATH, "encPath is null or empty");
        long h = MdixNative.loadEncrypted(encPath, keyPath != null ? keyPath : "");
        if (h == 0)
            throw new MdixException(MdixException.Kind.IO_ERROR, "failed to load encrypted: " + encPath);
        return new Database(h);
    }

    /**
     * Loads an encrypted .mdix.enc file using a password.
     */
    public static Database loadEncryptedPassword(String encPath, String password) {
        if (encPath == null || encPath.isEmpty())
            throw new MdixException(MdixException.Kind.INVALID_PATH, "encPath is null or empty");
        if (password == null || password.isEmpty())
            throw new MdixException(MdixException.Kind.INVALID_PATH, "password is null or empty");
        long h = MdixNative.loadEncryptedPassword(encPath, password);
        if (h == 0)
            throw new MdixException(MdixException.Kind.IO_ERROR,
                "failed to load encrypted (password): " + encPath);
        return new Database(h);
    }

    /** Shortcut: parses a JSON object string into a Database. */
    public static Database loadJson(String json) {
        return CONVERTER.fromJson(json);
    }

    /** Shortcut: parses a TOML table string into a Database. */
    public static Database loadToml(String toml) {
        return CONVERTER.fromToml(toml);
    }

    // ── Builder ───────────────────────────────────────────────────────────────

    /** Creates a new empty {@link Builder}. Remember to close it. */
    public static Builder newBuilder() {
        return new Builder();
    }

    // ── Converter ─────────────────────────────────────────────────────────────

    /** Returns the shared {@link Converter} instance. */
    public static Converter convert() {
        return CONVERTER;
    }
}

File: mdix-go/dixscript.go
// Package dixscript provides Go bindings for the DixScript (.mdix) runtime.
//
// Quick start:
//
//	db, err := dixscript.LoadStr(`@DATA( port = 8080, host = "localhost" )`)
//	if err != nil {
//	    log.Fatal(err)
//	}
//	defer db.Close()
//
//	port, _ := db.GetInt("port")
//	host, _ := db.GetString("host")
//	fmt.Printf("connecting to %s:%d\n", host, port)
//
// Build requirements:
//  1. CGO_ENABLED=1  (the default for native builds)
//  2. Run `cargo build -p mdix-ffi` to populate:
//     - internal/include/mdix_ffi.h
//     - internal/lib/<GOOS>-<GOARCH>/libmdix_ffi.*
package dixscript

import "github.com/Mid-D-Man/dixscript-go/internal"

// Version returns the DixScript native library version string.
func Version() string {
	return internal.Version()
}

// ── Loading ───────────────────────────────────────────────────────────────────

// Load loads a .mdix file from disk.
// The caller must close the returned Database.
func Load(path string) (*Database, error) {
	if path == "" {
		return nil, errNative("path is empty")
	}
	h := internal.Load(path)
	if h == nil {
		return nil, loadError(path)
	}
	return &Database{handle: h, sourcePath: path}, nil
}

// LoadStr loads .mdix content from a source string.
// The caller must close the returned Database.
func LoadStr(source string) (*Database, error) {
	if source == "" {
		return nil, errParse("source string is empty")
	}
	h := internal.LoadStr(source)
	if h == nil {
		return nil, parseError()
	}
	return &Database{handle: h}, nil
}

// LoadEncrypted loads an encrypted .mdix.enc file using a key file.
// Pass empty keyPath to auto-detect the key file next to the encrypted file.
// The caller must close the returned Database.
func LoadEncrypted(encPath, keyPath string) (*Database, error) {
	if encPath == "" {
		return nil, errNative("encPath is empty")
	}
	h := internal.LoadEncrypted(encPath, keyPath)
	if h == nil {
		return nil, loadError(encPath)
	}
	return &Database{handle: h}, nil
}

// LoadEncryptedPassword loads an encrypted .mdix.enc file using a password.
// The caller must close the returned Database.
func LoadEncryptedPassword(encPath, password string) (*Database, error) {
	if encPath == "" {
		return nil, errNative("encPath is empty")
	}
	if password == "" {
		return nil, errNative("password is empty")
	}
	h := internal.LoadEncryptedPassword(encPath, password)
	if h == nil {
		return nil, loadError(encPath)
	}
	return &Database{handle: h}, nil
}

// LoadEncryptedBytes loads encrypted data from a byte slice.
// keyContent is the full text content of the .mdix.key file.
// Pass empty password when using key-file mode.
// The caller must close the returned Database.
func LoadEncryptedBytes(data []byte, keyContent, password string) (*Database, error) {
	if len(data) == 0 {
		return nil, errNative("encrypted byte slice is empty")
	}
	if keyContent == "" {
		return nil, errNative("keyContent is empty")
	}
	h := internal.LoadEncryptedBytes(data, keyContent, password)
	if h == nil {
		return nil, errNative(internal.LastError())
	}
	return &Database{handle: h}, nil
}

// ── JSON/TOML shortcuts (delegates to Convert) ────────────────────────────────

// LoadJSON parses a JSON object string into a new Database.
// The caller must close the returned Database.
func LoadJSON(json string) (*Database, error) {
	return Convert.FromJSON(json)
}

// LoadToml parses a TOML table string into a new Database.
// The caller must close the returned Database.
func LoadToml(toml string) (*Database, error) {
	return Convert.FromToml(toml)
}

// ── Builder shortcut ──────────────────────────────────────────────────────────

// NewBuilder creates a new empty Builder.
// The caller must close the returned Builder.
func NewBuilderFunc() *Builder {
	return NewBuilder()
}

// ── Private helpers ───────────────────────────────────────────────────────────

func loadError(path string) error {
	if msg := internal.LastError(); msg != "" {
		return errIO(msg)
	}
	return errIO("failed to load: " + path)
}

func parseError() error {
	if msg := internal.LastError(); msg != "" {
		return errParse(msg)
	}
	return errParse("failed to parse source")
}
